### DISCUSSION TITLE
Does CPM work with CMakePresets?

### DISCUSSION THREAD

**Discussion by janos-ijgyarto** [cpm-cmake/CPM.cmake#493] [Q&A]
Hi all,

I have a somewhat complicated project setup, where I use [CMakePresets combined with VS 2022](https://learn.microsoft.com/en-us/cpp/build/cmake-presets-vs) to manage different configurations (compiler, generator, platform, etc.) Because of how CMakePresets currently works, practically every one of the variables will require a whole different config preset, and thus a whole new configuration. For builds, this was not a problem, I could just set it so CMake's output and the binaries would be placed in a subfolder named after the config preset. 

The problems began when I tried to combine this with [FetchContent](https://cmake.org/cmake/help/latest/module/FetchContent.html). Because of how it works, it wasn't possible to "share" the fetched libraries between different build configs. Ideally all I want is that I fetch the sources once, configure if needed (some libraries are header-only and don't even contain a CMakeLists.txt), and output the builds based on the current configuration (i.e make it behave as a package manager)

With how FetchContent works, however, I had to set the FetchContent base directory within the config-specific subfolder, which means I had to re-fetch every dependency for every different configuration (including when I just swap between debug and release). There is a [thread](https://discourse.cmake.org/t/fetchcontent-base-directory-and-cmakepresets/6722) on the CMake forums which illustrates the issue.

Does CPM offer a way around this? Or would I end up having the same issue?

**Comment by TheLartians** [comment]
Not sure I understand the problem 100%, but would setting the [`CPM_SOURCE_CACHE`](https://github.com/cpm-cmake/CPM.cmake#cpm_source_cache) environment variable solve the issue? That way any dependencies added with CPM will be downloaded to the source cache directory only once for all projects that share the same cache.

**Comment by audricschiltknecht** [comment]
Hello,

I've got the same initial question.
For me, that would allow to define a set of CMake variables in a preset of a library and abstract them for our users, who can then simply use the "user" preset, instead of having to manually specify each CMake variable as an option when adding a dependency via CPM.

**Comment by oceanofthelost** [comment]
@janos-ijgyarto, one way around having to have a separate configuration per compiler, generator, etc, is to use a multi config generator. Currently CMake supports two, `Ninja Multi-Config` and `Visual Studio`. 

When using a CMakePreset, build and test presents need to explicitly set `configuration` in order have the correct configuration built. 

With this setup, a single build step or workflow can be used and CPM will work as expected. 

If a multi-config generator cannot be used, then CPM cache will need to be set and shared across projects.
